contracts/data/DoublyLinkedList.sol

```diff
@@ -0,0 +1,428 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Doubly linked list implementation with enumeration functions
+ */
+library DoublyLinkedList {
+    struct DoublyLinkedListInternal {
+        mapping(bytes32 => bytes32) _nextValues;
+        mapping(bytes32 => bytes32) _prevValues;
+    }
+
+    struct Bytes32List {
+        DoublyLinkedListInternal _inner;
+    }
+
+    struct AddressList {
+        DoublyLinkedListInternal _inner;
+    }
+
+    struct Uint256List {
+        DoublyLinkedListInternal _inner;
+    }
+
+    error DoublyLinkedList__InvalidInput();
+    error DoublyLinkedList__InvalidReference();
+
+    function contains(
+        Bytes32List storage self,
+        bytes32 value
+    ) internal view returns (bool) {
+        return _contains(self._inner, value);
+    }
+
+    function contains(
+        AddressList storage self,
+        address value
+    ) internal view returns (bool) {
+        return _contains(self._inner, bytes32(uint256(uint160(value))));
+    }
+
+    function contains(
+        Uint256List storage self,
+        uint256 value
+    ) internal view returns (bool) {
+        return _contains(self._inner, bytes32(value));
+    }
+
+    function prev(
+        Bytes32List storage self,
+        bytes32 value
+    ) internal view returns (bytes32) {
+        return _prev(self._inner, value);
+    }
+
+    function prev(
+        AddressList storage self,
+        address value
+    ) internal view returns (address) {
+        return
+            address(
+                uint160(
+                    uint256(
+                        _prev(self._inner, bytes32(uint256(uint160(value))))
+                    )
+                )
+            );
```

Comment:
We convert the address to/from bytes32 a few times throughout SS. What do you think about adding a `toAddress` from `fromAddress` to the `AddressUtils`, or `SafeCast` library?

Reply:
I thought the same thing, but this might necessitate adding an `UnsafeCast` library because within the data structures we can be sure that we're not losing data by casting, and don't want to waste gas by checking.

I think we'll do this at some point, but not in this PR.